alchemy/src/sqlite/proxy.ts

```diff
@@ -0,0 +1,132 @@
+import type { Scope } from "../scope.ts";
+import type { State, StateStore } from "../state.ts";
+import { withExponentialBackoff } from "../util/retry.ts";
+
+/**
+ * Wraps a local or remote state store implementation
+ * and handles serialization/deserialization.
+ */
+export abstract class StateStoreProxy implements StateStore {
+  private dispatch?: Promise<StateStoreProxy.Dispatch>;
+
+  constructor(readonly scope: Scope) {}
+
+  abstract provision(): Promise<StateStoreProxy.Dispatch>;
+
+  private async run<TMethod extends StateStoreProxy.Method>(
+    method: TMethod,
+    params: StateStoreProxy.API[TMethod]["params"],
+  ): Promise<StateStoreProxy.API[TMethod]["result"]> {
+    this.dispatch ??= this.provision();
+    return this.dispatch.then(async (dispatch) => {
+      const result = await withExponentialBackoff(
+        () => dispatch(method, params),
+        () => true,
+      );
+      return result;
+    });
+  }
+
+  init(): Promise<void> {
+    return this.run("init", []);
+  }
```

Comment:
Couldn't we do this with an actual proxy instead of enumerating each of the methods?

Reply:
Great question! The reason why I did it this way is because the serialization logic depends on the method. For example:
- `set()` serializes the input value
- `get()` deserializes a single output value
- `all()` and `getBatch()` contain multiple output values that each need to be deserialized separately (or can they be deserialized together?)

Perhaps we can just run everything through serde instead so it can just be a proxy? The other thing is that I figured we'd want all the state stores to be classes (since the `new StateStore()` syntax is currently standard across the board), which (I think) wouldn't be possible if they were wrapping a proxy instead of inheriting another classes.

Reply:
Oh wow, good to know. Feels hacky but definitely can see the use case for that.